docs/adrs/0000-caching-dependencies.md

```diff
@@ -0,0 +1,59 @@
+# 0. Caching dependencies
+Date: 2021-05-21
+
+Status: Proposed
+
+# Context
+`actions/setup-node` is the 2nd most popular action in GitHub Actions. A lot of customers use it in conjunction with [actions/cache](https://github.com/actions/cache) to speed up dependencies installation.  
+See more examples on proper usage in [actions/cache documentation](https://github.com/actions/cache/blob/main/examples.md#node---npm).
+
+# Goals & Anti-Goals
+Integration of caching functionality into `actions/setup-node` action will bring the following benefits for action users:
+- Decrease the entry threshold for using the cache for Node.js dependencies and simplify initial configuration
+- Simplify YAML pipelines because no need additional steps to enable caching
+- More users will use cache for Node.js so more customers will have fast builds!
+
+As the first stage, we will add support for NPM dependencies caching. We can consider adding the same functionality for Yarn later.
+
+We don't persue the goal to provide wide customization of caching in scope of `actions/setup-node` action. The purpose of this integration is covering ~90% of basic use-cases. If user needs flexible customization, we should advice them to use `actions/cache` directly.
+
+# Decision
+- Add `cache` input parameter to `actions/setup-node`. For now, input will accept the following values: 
```

Comment:
We have thought about making this input boolean like `cache: true` and it would work for us for now. But if we want to extend caching to Yarn in future, we will need to add additional field to choose between Yarn / npm. Npm 7 started t support `yarn.lock` so we can't choose package manager by filename anymore.

Any thoughts and feedback are appreciated

Reply:
I really want this to be boolean. can we not automatically detect the package manager?

Reply:
No way unfortunately, because we can't rely on file name.
I agree on having `cache: boolean` but in this case, we will need additional field in future like `package-manager`

Reply:
I think `cache: npm` is the best name that we can then re-use across some of our other first party actions.

For example long-term with `setup-python`, there are a bunch of package managers like `pip`, `conda` and `virtualenv` that would have to be supported so having `cache: true/false` would be very hard to extend

Reply:
That is really good point